Provide initial functionality

dav3r opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR contains the initial functionality for this Lambda, which is to scan a specified set of AWS accounts and publish files/objects (to an S3 bucket) containing the public IP addresses of EC2 instances or Elastic IPs that have been properly tagged.

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

We need a mechanism to publish egress IP info from a subset of our AWS accounts and this Lambda will allow us to do just that.

This is part of the work for:
* https://github.com/cisagov/cool-system/issues/208
* https://github.com/cisagov/cyhy_amis/issues/284

This PR is related to the following:
* https://github.com/cisagov/cool-accounts/pull/126
* https://github.com/cisagov/cool-accounts/pull/127

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

In addition to passing all of the automated pre-commit checks, this code was successfully built and tested both manually (using the "Test" tab on the AWS Lambda web console) and in conjunction with the Terraform in the [`cisagov/publish-egress-ip-terraform`](https://github.com/cisagov/publish-egress-ip-terraform) repository.

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##
- [x] Re-test to ensure Lambda is functioning as expected after all code changes during review.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [x] Create a release.

dav3r committed:
Remove unneeded content inherited from skeleton

dav3r committed:
Customize Dockerfile and docker-compose.yml for this project

dav3r committed:
Update Pipfiles for this project

dav3r committed:
Update Pipfile.lock files for each supported Python version

These files are created by running "pipenv lock" in each directory.

dav3r committed:
Add boto3 to requirements.txt

It is needed to make various AWS API calls in the Lamba code.

dav3r committed:
Replace skeleton handler with code to publish egress IPs

dav3r committed:
Replace skeleton text with useful documentation

dav3r committed:
Bump version from 0.0.1 to 1.0.0

mcdonnnj commented on requirements.txt:
```diff
@@ -1,2 +1,3 @@
+boto3
```
I don't think this is necessary to build the Lambda.

```suggestion
```

dav3r replied:
Are you saying that `boto3` instead belongs in `requirements-dev.txt` (or `requirements-test.txt`) and not in `requirements.txt`?  I need to ensure `boto3` is installed somewhere in my pyenv for local testing.  Or are you saying it doesn't belong in any of those places?

mcdonnnj replied:
Why do you need `boto3` in the (non-Lambda) Python environment?

jsf9k replied:
Often tasks such as pushing to or pulling from S3 are delegated to `localhost`, so that the user's AWS creds (when testing locally) or GH Actions' (when testing in GH Actions) can be used.  I assume that is what is happening here.

dav3r replied:
I was mainly thinking about when I had tested portions of the lamdba code locally and I needed to make AWS calls.  Is everyone cool with me just moving `boto3` from `requirements.txt` into `requirements-dev.txt`?

mcdonnnj replied:
I guess my question is do we need `boto3` for continued, regular development of this project? Should future development efforts expect to need `boto3` to work on this project?

dav3r replied:
I contend that it's handy to include it and doesn't really hurt anything, but I will gladly follow the will of the people here.

mcdonnnj replied:
I guess my thinking is you could spin up the Docker image created with this configuration for testing to mimic the environment it would run in when deployed.

jsf9k replied:
I'm fine with it going into `requirements-dev.txt` if a comment is added there explaining why it is being added.

dav3r replied:
> I'm fine with it going into `requirements-dev.txt` if a comment is added there explaining why it is being added.

@mcdonnnj Does that work for you?

mcdonnnj replied:
Yes with an explanatory comment that is fine for me.

dav3r replied:
Have a look at https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/6de01f31e5a227cfd0937ec8cc988f96b63abc64 and let me know if that works for you.

dav3r commented:
@jsf9k or @mcdonnnj - Any thoughts on adding `publish_egress_ip.zip` (the output of `docker compose up build_deployment_package`) to the `.gitignore` file?  This is something that probably should be decided over in [`cisagov/skeleton-aws-lambda-python`](https://github.com/cisagov/skeleton-aws-lambda-python), but since I have this PR open now, I figured I'd ask here.

jsf9k commented:
> @jsf9k or @mcdonnnj - Any thoughts on adding `publish_egress_ip.zip` (the output of `docker compose up build_deployment_package`) to the `.gitignore` file? This is something that probably should be decided over in [`cisagov/skeleton-aws-lambda-python`](https://github.com/cisagov/skeleton-aws-lambda-python), but since I have this PR open now, I figured I'd ask here.

Sounds good to me.  I agree that it should be in the skeleton.

dav3r committed:
Add gitignore entry for default lambda deployment package name

dav3r commented:
> > @jsf9k or @mcdonnnj - Any thoughts on adding `publish_egress_ip.zip` (the output of `docker compose up build_deployment_package`) to the `.gitignore` file? This is something that probably should be decided over in [`cisagov/skeleton-aws-lambda-python`](https://github.com/cisagov/skeleton-aws-lambda-python), but since I have this PR open now, I figured I'd ask here.
> 
> Sounds good to me. I agree that it should be in the skeleton.

I realized that we already do this [elsewhere](https://github.com/cisagov/findings-data-import-lambda/blob/239524883a07063d3d904cf3b4f6902ad9fbae7d/.gitignore#L5-L6) and I'm not doing anything new here, so I added the `.gitignore` entry in https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/0d2eca32e437e70d24da561d49781a878503e385.

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -1,22 +1,142 @@
+) -> boto3.client:
+    """Assume the given role and return an AWS client for the given service using that role."""
+    # Create an STS session with current credentials
+    sts: boto3.client = boto3.client("sts")
+
+    # Assume the provided role
+    response: Dict[str, Any] = sts.assume_role(
+        RoleArn=role_arn, RoleSessionName=session_name
+    )
```
I think this logic should be broken out into a separate function that returns a tuple with the necessary session/access key items since this logic already duplicated in `create_assumed_aws_resource()`.

dav3r replied:
Good thinking- I added that in https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/96951ae9430f155015a946758a4cf60d58dd55bb, let me know what you think.  

Doh, just realized that I didn't give you co-commit props, sorry about that!

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -1,22 +1,142 @@
+            tags: Dict[str, str] = {x["Key"]: x["Value"] for x in instance.tags}
+        except TypeError:
+            # This happens if there are no tags associated with the instance
+            tags = {}
```
Another case where I think a helper function would make this a bit cleaner because of duplication.

dav3r replied:
Heck yeah - see https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/7e5b01aeebaa7c0257a13a056f97d52b149e3ba8 for this.

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -34,54 +154,158 @@ def task_default(event):
+    # {description} - description of the published file
+    file_header: str = event.get(
+        "file_header",
+        "###\n# https://{domain}/{filename}\n# {timestamp}\n# {description}\n###\n",
```
This is just a thought but would this be better represented as a `list(string)` of lines that get joined? That way it's a bit clearer what each piece of the header is when creating/reading for configuration. This could be future work if it seems reasonable.

dav3r replied:
This is a nice idea, which I implemented in https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/d27545d3653cce8d2c58987ba07baf42e2ae336d.

I now have to make corresponding changes over in https://github.com/cisagov/publish-egress-ip-terraform/pull/6, so let's not close this conversation until I have confirmed that is done.

dav3r replied:
The corresponding Terraform change has been made over in https://github.com/cisagov/publish-egress-ip-terraform/pull/6/commits/d1c0a2906d3996c536eac8d198d23efc7175c303.

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -34,54 +154,158 @@ def task_default(event):
+
+    # Initialize application regexes and a set to accumulate IPs for each file
+    for config in file_configs:
+        config["app_regex"] = re.compile(config["app_regex"])
+        config["ip_set"] = {ip_network(i) for i in config["static_ips"]}
```
There are no guarantees I am aware of that these fields exist. The `FileConfig` class just provides an easy wrapper for static analysis. Per the `TypedDict` description:
> `TypedDict` declares a dictionary type that expects all of its instances to have a certain set of keys, where each key is associated with a value of a consistent type. This expectation is not checked at runtime but is only enforced by type checkers.

We would need to use something like [Schema](https://github.com/keleshev/schema) to actually validate this user input. Another option is to use a [dataclass](https://docs.python.org/3/library/dataclasses.html) and just pass the config entry with `**kwargs` to initialize a given object. We could define a [`__post_init__()` method](https://docs.python.org/3/library/dataclasses.html#post-init-processing) to handle things like parsing the static IPs and compiling the regex.

dav3r replied:
I kept it simple and added some validation in https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/d6d38205f7ea56f3f95c65b323434b25bfa0829e.  Let me know if you think there's a need to go beyond that.

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -34,54 +154,158 @@ def task_default(event):
+        config["app_regex"] = re.compile(config["app_regex"])
+        config["ip_set"] = {ip_network(i) for i in config["static_ips"]}
+
+    # Name of the AWS tag whose value represents the application
```
Mirroring my suggestion in the Terraform implementation.

```suggestion
    # Name of the AWS resource tag whose value represents the application
```

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -34,54 +154,158 @@ def task_default(event):
+    # Name of the AWS tag whose value represents the application
+    # associated with an IP address
+    application_tag_name: str = event.get("application_tag", "Application")
+    # AWS tag name indicating whether an IP address should be published
```
Ditto.

```suggestion
    # AWS resource tag name indicating whether an IP address should be published
```

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -34,54 +154,158 @@ def task_default(event):
+
+    # Update each file in the bucket
+    for config in file_configs:
+        # Initialize bucket contents
+        bucket_contents = file_header
```
I am not a fan of `bucket_contents` as a variable name for what's being done here. I think this is file/object content creation and although it goes into the bucket it does not necessarily represent all of the contents of the bucket.

dav3r replied:
Agreed- that nomenclature is old and left over from the original code, so I cleaned it up in https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/327ddef3a7f0dd29e44eb62289096b075dc19ae3.  Let me know if you have any feedback.

dav3r committed:
Improve specificity of comments

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
DRY out code by adding an assume_role() function

dav3r committed:
DRY out code by adding a convert_tags() function

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Improve variable names related to S3 objects

Also, update comments appropriately.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Add some additional type hints

dav3r committed:
Validate user-supplied file_configs data

Log any errors if necessary.

mcdonnnj commented on Dockerfile:
```diff
@@ -28,9 +28,9 @@ RUN yum update --assumeyes \
-    pip \
-    setuptools \
-    wheel \
+  pip \
+  setuptools \
+  wheel \
```
Did the spacing on these need to be changed?

```suggestion
    pip \
    setuptools \
    wheel \
```

dav3r replied:
That is how VS Code autoformatted it and I actually agree here.  I think `skeleton-aws-lambda-python` should be modified to use this spacing, which aligns with the rest of the Dockerfile (i.e. continued lines are indented two spaces).  I don't see why these particular lines (and only these) should be indented four spaces.

mcdonnnj replied:
They are indented four spaces to separate that they are arguments to a command since this Docker instruction chains commands:
https://github.com/cisagov/publish-egress-ip-lambda/blob/ac3ddf120769e0932628fceacf7b172329599c4c/Dockerfile#L30-L34
Just as a refresher it chains commands to ensure that `pipenv` (and its dependencies) have the latest versions of `pip`, `setuptools`, and `wheel` available when installing.

mcdonnnj replied:
As a thought the syntax is borrowed from more traditional Docker projects that pin these so we could just do
```dockerfile
RUN python3 -m pip install --no-cache-dir pip setuptools wheel
  && python3 -m pip install --no-cache-dir pipenv
```
here and in the skeleton to negate the issue entirely.

dav3r replied:
> As a thought the syntax is borrowed from more traditional Docker projects that pin these so we could just do
> 
> ```dockerfile
> RUN python3 -m pip install --no-cache-dir pip setuptools wheel
>   && python3 -m pip install --no-cache-dir pipenv
> ```
> 
> here and in the skeleton to negate the issue entirely.

I like this solution the best- see https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/29fbe09a56affdeac2aca9eeb88f3b01e5f43152.

mcdonnnj commented on README.md:
```diff
@@ -32,31 +31,64 @@ docker compose up build_deployment_package
+| bucket\_name | The name of the S3 bucket to publish egress IP address information to. | `string` | n/a | yes |
+| domain | The domain hosting the published file(s) containing egress IPs. | `string` | n/a | yes |
+| ec2\_read\_role\_name | The name of the IAM role that allows read access to the necessary EC2 attributes.  Note that this role must exist in each account that you want to query and it must be assumable by Lambda.  For an example policy, see [`cisagov/cool-accounts`](https://github.com/cisagov/cool-accounts/blob/develop/dynamic/ec2readonly_policy.tf). | `string` | `"EC2ReadOnly"` | no |
+| file\_configs | A list of objects that define the files to be published.  "app_regex" specifies a regular expression that is matched against the value of the application_tag to determine if the address should be included in the file.  "description" is the description of the published file.  "filename" is the name to assign the published file.  "static\_ips" is the list of CIDR blocks that will  ...
```
I know this is copied from the Terraform project but I would prefer we be a little more Pythonic in our terminology (I'm not going to harass you to rewrite this using Python type hint syntax).

```suggestion
| file\_configs | A list of dictionaries that define the files to be published.  "app_regex" specifies a regular expression that is matched against the value of the application_tag to determine if the address should be included in the file.  "description" is the description of the published file.  "filename" is the name to assign the published file.  "static\_ips" is the list of CIDR blocks that will always be included in the published file. | `list(dict({ app_regex = string, description = string, filename = string, static_ips = list(string) }))` | n/a | yes |
```

mcdonnnj commented on README.md:
```diff
@@ -32,31 +31,64 @@ docker compose up build_deployment_package
+| file\_configs | A list of objects that define the files to be published.  "app_regex" specifies a regular expression that is matched against the value of the application_tag to determine if the address should be included in the file.  "description" is the description of the published file.  "filename" is the name to assign the published file.  "static\_ips" is the list of CIDR blocks that will  ...
+| file\_header | The header template for each published file.  The following variables are available within the template: `{domain}` - the domain where the published files are located, `{filename}` - the name of the published file, `{timestamp}` - the timestamp when the file was published, `{description}` - the description of the published file. | `string` | `"###\n# https://{domain}/{filename}\n ...
+| publish\_egress\_tag | The name of the AWS tag whose value represents whether the EC2 instance or elastic IP should have its public IP address published. | `string` | `"Publish Egress"` | no |
+| region\_filters | A list of AWS EC2 region filters to use when querying for IP addresses to publish.  If a filter is not specified, the query will be performed in all regions.  An example filter to restrict to US regions looks like this: `[{ "Name" : "endpoint", "Values" : ["*.us-*"] }]`.  For more information, refer to the [AWS EC2 CLI documentation](https://docs.aws.amazon.com/cli/latest/refe ...
```
Same basis.

```suggestion
| region\_filters | A list of AWS EC2 region filters to use when querying for IP addresses to publish.  If a filter is not specified, the query will be performed in all regions.  An example filter to restrict to US regions looks like this: `[{ "Name" : "endpoint", "Values" : ["*.us-*"] }]`.  For more information, refer to the [AWS EC2 CLI documentation](https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-regions.html). | `list(dict({ Name = string, Values = list(string) }))` | `[]` | no |
```

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -1,22 +1,152 @@
+        yield (tags.get(application_tag_name, ""), instance.public_ip_address)
+
+    for vpc_address in vpc_addresses:
+        # Convert elastic IP tags from AWS dictionary into a Python dictionary
```
Consistent commenting.

```suggestion
        # Convert elastic IP tags from an AWS dictionary into a Python dictionary
```

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -34,54 +164,176 @@ def task_default(event):
+            return result
+
+        # Initialize the set of static IPs
+        config["ip_set"] = {ip_network(i) for i in config.get("static_ips", "")}
```
Since this is supposed to be a `list(string)` value.

```suggestion
        config["ip_set"] = {ip_network(i) for i in config.get("static_ips", [])}
```

dav3r replied:
Whoops, yes, good catch!

mcdonnnj reviewed (commented):
Some additional feedback based on changes that have been made (and some documentation items that I missed).

dav3r committed:
Use correct default value for a missing optional field

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Describe inputs more Pythonically

...and less Terraformally ;)

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Use consistent comment verbiage

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Make file_header a list of strings, instead of a string

When the file is published, newline characters are added between each item in the list.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -34,54 +164,194 @@ def task_default(event):
+        ],
+    )
+
+    # Ensure file_header is a list of strings
```
Should we also do something like:

```python
file_header = [str(e) for e in file_header]
```

to ensure that all elements are strings? If something is not the later `join()` will fail with a `TypeError`.

dav3r replied:
Yes, that is what I had originally intended, but your version is superior to what I hastily wrote; improved in https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/8e18d4d4119a232d6dfc2e02a777feadb6d04ec5.

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -34,54 +164,194 @@ def task_default(event):
+    # Update each object (file) in the bucket
+    for config in file_configs:
+        # Initialize contents of object to be published
+        object_contents = "\n".join(file_header)
```
With my proposed change we lost a `\n` at the end of the header.

```suggestion
        object_contents = "\n".join(file_header) + "\n"
```

mcdonnnj reviewed (commented):
Thank you for implementing that change to the file header format. I have a question and a suggestion related to that change.

dav3r committed:
Ensure file header ends with newline

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Move boto3 from requirements.txt to requirements-dev.txt

boto3 is not strictly required, but it can be useful for some local development testing activities, so we include it as a convenience.

Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>
Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Improve check for string elements in file header

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r commented:
@mcdonnnj - Can you please circle back to this review when you have a chance?

mcdonnnj commented on Dockerfile:
```diff
@@ -5,12 +5,12 @@ FROM amazon/aws-lambda-python:$PY_VERSION
 ARG PY_VERSION=3.9
 # Get the base output file name from the command line if provided.
-ARG FILE_NAME=example
+ARG FILE_NAME=publish_egress_ip
```
I'm looking at making this non-project specific in the skeleton but leaving it as an argument to allow for customized release workflows etc. as an option. Thoughts?

```suggestion
ARG FILE_NAME=lambda_build
```

dav3r replied:
Good idea- I like that plan.  Are you ok with me leaving this as is for now and making that change later when it gets added to the skeleton?  Until we move to automated releases/deployments, I'd prefer to avoid having generic names for these zip files because it could lead to confusion.

mcdonnnj replied:
Would you elaborate on why you think it could lead to confusion? I only ask because the `build` job already provides a different name in https://github.com/cisagov/publish-egress-ip-lambda/blob/35e58a7ac13f0d1805ae93415ce31d063b8a2239/.github/workflows/build.yml#L160 if you pull down the automatically build images.

dav3r replied:
> Would you elaborate on why you think it could lead to confusion?

If someone was building it locally and not overriding the `FILE_NAME` arg because they were being lazy. 😄
But I suppose my laziness is not a great reason, so I will make the change you suggested.

mcdonnnj replied:
Just a note for future laziness you could always use a `docker-compose.override.yml` and add
```yaml
services:
  build_deployment_package:
    build:
      args:
        FILE_NAME: publish_egress_ip
```
there x)

mcdonnnj commented on .gitignore:
```diff
@@ -2,6 +2,9 @@
 # See: https://git-scm.com/docs/gitignore
 
+## Project Specific ##
+publish_egress_ip.zip
```
In concert with the previous suggestion.

```suggestion
lambda_build.zip
```

mcdonnnj commented on README.md:
```diff
@@ -2,13 +2,12 @@
+This repository contains code to create a Lambda function that can scan a set
+of AWS accounts and publish files (to an S3 bucket) containing the public IP
+addresses of EC2 instances or Elastic IPs that have been properly tagged.
+Refer to the "Lambda inputs" section below, specifically the
```
I'm pretty sure this is the correct syntax.

```suggestion
Refer to the [Lambda inputs](#lambda-inputs) section below, specifically the
```

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -34,54 +164,196 @@ def task_default(event):
+    # in each AWS account. Note that this role must exist in each account.
+    ec2_read_role_name: str = event.get("role_name", "EC2ReadOnly")
+
+    account_ids: List[str] = event.get("account_ids", None)
```
Update this to respect that it should be of type `List[str]` and copying what we do for the `file_header` key earlier in this function.

```suggestion
    account_ids: List[str] = event.get("account_ids", [])

    # Ensure account_ids is a list of strings
    try:
        if type(account_ids) is not list:
            account_ids = [str(account_ids)]
        else:
            account_ids = [str(e) for e in account_ids]
    except Exception:
        error_msg = "account_ids must be a list of strings."
        logging.error(error_msg)
        failed_task(result, error_msg)
        return result

```

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -34,54 +164,196 @@ def task_default(event):
+    account_ids: List[str] = event.get("account_ids", None)
+    for account_id in account_ids:
+        # Verify AWS account ID is 12 digits
+        if not re.match(r"^\d{12}$", str(account_id)):
```
Since we've already converted these to strings with the above suggestion.

```suggestion
        if not re.match(r"^\d{12}$", account_id):
```

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -34,54 +164,196 @@ def task_default(event):
+            logging.error(error_msg, account_id)
+            failed_task(result, error_msg % account_id)
+
+        logging.info("Examining account: %s" % account_id)
```
```suggestion
        logging.info("Examining account: %s", account_id)
```

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -34,54 +164,196 @@ def task_default(event):
+        # Get a list of all regions that match our filter
+        regions: List[str] = get_ec2_regions(ec2, region_filters)
+
+        logging.info("Gathering public IPs from %d regions" % len(regions))
```
```suggestion
        logging.info("Gathering public IPs from %d regions", len(regions))
```

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -34,54 +164,196 @@ def task_default(event):
+
+        # Loop through the region list and fetch the public EC2 IPs
+        for region in regions:
+            logging.info("Querying region: %s" % region)
```
```suggestion
            logging.info("Querying region: %s", region)
```

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -34,54 +164,196 @@ def task_default(event):
+        if not re.match(r"^\d{12}$", str(account_id)):
+            error_msg = 'Account ID "%s" is invalid - it must be 12 digits.'
+            logging.error(error_msg, account_id)
+            failed_task(result, error_msg % account_id)
```
Should we return after this? Skip to the next account? I don't think we should just continue on if there's an issue with the account ID. A note if we skip is that any other account ID failures will overwrite the error message set by `failed_task()`.

dav3r replied:
My original thinking was to just have it go on to the next account, but I'm fine changing it to bail out (return) as soon as it hits an invalid account ID.

mcdonnnj replied:
I'm fine with either. My only concern was continuing to work on an account that is invalid so whichever way you go will solve that 👍

dav3r replied:
Remedied with https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/bba50bbc816877cc8434cef1873c9ac574ef03cd.

mcdonnnj replied:
Not to complicate things but thoughts on just making the account ID check a pre-flight thing like the bucket name? Optionally could build a list of invalid account IDs and `join()` them into the error message.

dav3r replied:
Dang it, you're right - that is a superior solution.  I'll do that and let you know here when it's done.

dav3r replied:
In https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/14c453181af74af74ccb6ba8e61b2a88e99ca92d, I check the account IDs prior to the main processing loop.  Arg, sorry @mcdonnnj - I forgot to give you co-commit credit, but thanks for the excellent suggestion.

As the number of pre-flight checks has increased, I decided to move all of them out of `task_publish()` and into a new function `validate_event_data()` with https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/f66553f29c61aaa7c0e7657ffab32a7f850e5ae8.

dav3r replied:
@mcdonnnj - Can this conversation be resolved?

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -34,54 +164,196 @@ def task_default(event):
+    now = "{:%a %b %d %H:%M:%S UTC %Y}".format(datetime.utcnow())
+
+    # The bucket to publish the files to
+    bucket_name: str = event["bucket_name"]
```
Since this will fail if a value is not provided I think we should check this earlier in the function to bail out before processing everything if this is missing.

dav3r replied:
Good thinking!  Done in https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/16c05b94a2cd54c16e59bd61c2933862cc54cae5, along with some error handling and a type check.

mcdonnnj replied:
Beautiful, thanks!

dav3r committed:
Default to empty list if account_ids is missing

Also, use same kind of checks and error handling as we do with file_header key.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Replace a section mention with a Markdown anchor link

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Remove an obsolete string conversion

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Use our preferred logging syntax

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Check bucket_name key existence earlier in the process

While we are at it, add error handling and a  type check on the value.  This allows us to fail sooner if the bucket name was not supplied or isn't a string.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Bail out entirely if an invalid AWS account ID is encountered

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Validate all AWS account ID formats earlier

Now, when we start iterating through accounts to look for egress IPs, every account ID has been verified to be a 12-digit string.

dav3r committed:
Move all event data validation into a new function

dav3r committed:
Alphabetize variable initializations

dav3r commented:
@jsf9k - I re-requested your review since there has been a decent amount of change since your first review.

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -34,54 +164,253 @@ def task_default(event):
-    return result
+            # Verify that each provided AWS account ID is 12 digits
+            invalid_account_ids = []
+            for account_id in account_ids.copy():
+                if not re.match(r"^\d{12}$", account_id):
+                    account_ids.remove(account_id)
```
What is the rationale for copying the list and removing elements from the original?

dav3r replied:
I make a copy because removing one or more items from a list while iterating through that same list will skip items in the list.  As with most things in Python, there are other ways to accomplish this- see [here](https://sopython.com/canon/95/removing-items-from-a-list-while-iterating-over-the-list/) for a nice short post on this subject.

I suppose I could have also built up a new list of `valid_account_ids` and just left the original list of account IDs alone.

Did you have a different solution in mind or would you rather see an explanatory comment here?

mcdonnnj replied:
Sorry that was a poorly worded question on my part. I'm familiar with the need to copy a list to remove elements from it while iterating over it in Python. The core of my inquiry was why are we going through the trouble of removing failing account IDs from the list of account IDs.

dav3r replied:
Ah, now I understand.  That logic was left over from when I had previously planned to weed out bad account IDs and continue processing the valid accounts.  Now that we bail out if any errors are found in the event data, it no longer makes sense to remove the bad IDs from the list - updated in https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/a9fa4909a5d0443e60876d14f17ecf264add619e.

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -1,22 +1,152 @@
+
+
+def get_ec2_regions(
+    ec2: boto3.client, filter: Optional[List[Dict[str, Union[str, List[str]]]]] = []
+) -> List[str]:
+    """Get a filtered list of all the regions with EC2 support."""
+    response = ec2.describe_regions(Filters=filter)
```
There is a Python built-in `filter()` so we should avoid redefining it just as a best practices thing. Additionally, and this behavior was not something I learned until recently, we should avoid defining data structures as default values as they are reused for all invocations of the function. In this case it does not matter but if we were to make changes to the default list that change would propagate across invocations.

```suggestion
    ec2: boto3.client, filters: Optional[List[Dict[str, Union[str, List[str]]]]] = None
) -> List[str]:
    """Get a filtered list of all the regions with EC2 support."""
    if filters is None:
        filters = []

    response = ec2.describe_regions(Filters=filters)
```

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -34,54 +164,253 @@ def task_default(event):
+        account_ids: List[str] = event["account_ids"]
+        try:
+            # Ensure account_ids is a list of strings
+            if type(account_ids) is not list:
```
So this actually came up when I was using `pylint` to clean up another project. `isinstance()` is preferred vs `type()` because it is both faster and considers object inheritance. So this would validate not just a `list` but also if it is a subclass of `list` whereas `type()` only checks if it is a `list`. We should only use `type()` if we want to check if something is a specific class and not a subclass. Not particularly impactful here but a good best practice to get into.

```suggestion
            if not isinstance(account_ids, list):
```

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -34,54 +164,253 @@ def task_default(event):
+    # Bucket name checks
+    if "bucket_name" not in event:
+        errors.append('Missing required key "bucket_name" in event.')
+    elif type(event["bucket_name"]) != str:
```
Same comment about `isinstance()` vs `type()`.

```suggestion
    elif not isinstance(event["bucket_name"], str):
```

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -34,54 +164,253 @@ def task_default(event):
-            logging.error(error_msg)
-            logging.error(variable_error_msg, numerator, denominator)
-            failed_task(result, error_msg)
+            if type(file_header) is not list:
```
Same comment about `isinstance()` vs `type()`.

```suggestion
            if not isinstance(file_header, list):
```

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -34,54 +164,253 @@ def task_default(event):
+    # Account ID checks
+    if "account_ids" not in event:
+        errors.append('Missing required key "account_ids" in event.')
+    elif event["account_ids"] == []:
```
I personally prefer the truthy check for empty lists.

```suggestion
    elif not event["account_ids"]:
```

mcdonnnj reviewed (approved):
Better than ever! I had some best practices suggestions and a question for your consideration.

mcdonnnj commented on README.md:
````diff
@@ -32,31 +31,64 @@ docker compose up build_deployment_package
+
+```json
+{
+    "accounts": [
````
```suggestion
    "account_ids": [
```

mcdonnnj commented on README.md:
```diff
@@ -32,31 +31,64 @@ docker compose up build_deployment_package
+| account\_ids | The list of AWS accounts to query for egress IPs to publish. | `list(string)` | n/a | yes |
+| application\_tag | The name of the AWS tag whose value represents the application associated with an IP address. | `string` | `"Application"` | no |
+| bucket\_name | The name of the S3 bucket to publish egress IP address information to. | `string` | n/a | yes |
+| domain | The domain hosting the published file(s) containing egress IPs. | `string` | n/a | yes |
```
The Lambda uses a `.get()` with a default for this value.

```suggestion
| domain | The domain hosting the published file(s) containing egress IPs. | `string` | `"example.gov"` | no |
```

mcdonnnj commented on README.md:
```diff
@@ -32,31 +31,64 @@ docker compose up build_deployment_package
+        "123456789012",
+        "234567890123"
+    ],
+    "bucket": "my-egress-ip-bucket",
```
```suggestion
    "bucket_name": "my-egress-ip-bucket",
```

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -34,54 +164,253 @@ def task_default(event):
-        failed_task(result, error_msg)
-    else:
+    # File configuration checks
+    file_configs: List[FileConfig] = event["file_configs"]
```
Should we lock this stuff behind a `if "file_configs" in event` check like we do for `file_header` below? If not we should use a `.get()` since this is not a required key (and we don't want to bomb out with a `KeyError`).

dav3r replied:
Good catch- I overlooked this when I made the changes in https://github.com/cisagov/publish-egress-ip-lambda/commit/f66553f29c61aaa7c0e7657ffab32a7f850e5ae8.  Let me know if you have any concerns with what I did in https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/7f14b59727832b479d4b1f9d7ab0e0c15a50c72d to address this.

mcdonnnj reviewed (commented):
![image](https://user-images.githubusercontent.com/50747025/189390821-29b14eed-1fe4-4491-875e-54ea69643c39.png)

dav3r committed:
Avoid redefining built-in filter() function

Also, replace a mutable default argument value (empty list) with None and instead assign it inside the function.  For a good explanation of why we do this (and you should too), see: https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Prefer isinstance() over type()

This is a good best practice since isinstance() is faster and it considers object inheritance, which is generally preferable.

Want more info?  Here's a nice blog post on this subject: https://switowski.com/blog/type-vs-isinstance

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Correct documentation regarding a default value

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Correct key names in example input JSON

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Prefer truthy check for empty list over equality check

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Ensure file_configs gets an appropriate default value

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Simplify logic when validating account IDs

Now that we bail out if any errors are found in the event data, there is no longer any need to remove bad account IDs from the account_ids list.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Remove an unneeded entry from .gitignore

dav3r committed:
Customize Lambda deployment package BUILD_FILE_NAME in build job

dav3r committed:
Explain in README how to manually customize the Lambda deployment filename

dav3r committed:
Reformat pipfiles to satisfy linter

dav3r committed:
Revert "Customize Lambda deployment package BUILD_FILE_NAME in build job"

This reverts commit b6a91d15c4e925d466ee2bd8cb11aea4d5b1c7b4.

jsf9k commented on README.md:
```diff
@@ -72,12 +53,64 @@ cd src/py3.9
+| file\_header | The header template for each published file, comprised of a list of strings.  When the file is published, newline characters are automatically added between each item in the list.  The following variables are available within the template: `{domain}` - the domain where the published files are located, `{filename}` - the name of the published file, `{timestamp}` - the timestamp wh ...
+| publish\_egress\_tag | The name of the AWS tag whose value represents whether the EC2 instance or elastic IP should have its public IP address published. | `string` | `"Publish Egress"` | no |
+| region\_filters | A list of AWS EC2 region filters to use when querying for IP addresses to publish.  If a filter is not specified, the query will be performed in all regions.  An example filter to restrict to US regions looks like this: `[{ "Name" : "endpoint", "Values" : ["*.us-*"] }]`.  For more information, refer to the [AWS EC2 CLI documentation](https://docs.aws.amazon.com/cli/latest/refe ...
+| task | The name of Lambda task to perform.  Currently, the only valid value is `"publish"`. | `string` | n/a | yes |
```
```suggestion
| task | The name of the Lambda task to perform.  Currently, the only valid value is `"publish"`. | `string` | n/a | yes |
```

jsf9k commented on src/lambda_handler.py:
```diff
@@ -1,24 +1,157 @@
 
 
-def failed_task(result: dict[str, Any], error_msg: str) -> None:
+def assume_role(role_arn: str, session_name: str) -> Tuple[str, str, str]:
```
Would it be better to use a dictionary or [named tuple](https://docs.python.org/3/library/collections.html#collections.namedtuple) here?

dav3r replied:
Good suggestion, thanks!  Please see https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/a9ec40caa6dbf52a8c2f10e9afd6942702c52de4.

jsf9k commented on src/lambda_handler.py:
```diff
@@ -1,24 +1,157 @@
+
+def get_ec2_ips(
+    ec2: boto3.resource, application_tag_name: str, publish_egress_tag_name: str
+) -> Iterator[Tuple[str, str]]:
```
Again, I think a dictionary or named tuple might be better here.

dav3r replied:
Please see https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/a9ec40caa6dbf52a8c2f10e9afd6942702c52de4.

jsf9k commented on src/lambda_handler.py:
```diff
@@ -36,54 +169,252 @@ def task_default(event):
+                )
+
+        # Initialize the set of static IPs in this config
+        config["ip_set"] = {ip_network(i) for i in config.get("static_ips", [])}
```
Does this throw something you can catch if the static IPs are incorrectly formatted?

dav3r replied:
Indeed we can; that is an excellent idea.  Please have a look at https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/a7e839cfbfb667606c14cf56ccd90ed68e25b650.

dav3r replied:
PS: I tested this by deploying the lambda and including an invalid static IP address.  I confirmed that it correctly logged an error as intended.

jsf9k commented on src/lambda_handler.py:
```diff
@@ -36,54 +169,252 @@ def task_default(event):
+            else:
+                file_header = [str(e) for e in file_header]
+        except Exception:
+            errors.append("file_header must be a list of strings.")
```
I think `file_header` should be double-quoted here for consistency.

dav3r replied:
Is this what you mean?
```suggestion
            errors.append('"file_header" must be a list of strings.')
```

jsf9k replied:
Yes, that is what I mean.

dav3r replied:
Please see https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/bfb38499afdb29ac622f687a364088e8bf19e119.

jsf9k commented on src/lambda_handler.py:
```diff
@@ -36,54 +169,252 @@ def task_default(event):
+
+def validate_event_data(
+    event: Dict[str, Any]
+) -> Tuple[Dict[str, Any], bool, List[str]]:
```
Again, I suggest using a dictionary or named tuple here.

dav3r replied:
Please see https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/a9ec40caa6dbf52a8c2f10e9afd6942702c52de4.

jsf9k reviewed (approved):
Great work!  I have some comments and suggestions, but nothing that should hold up this PR.

dav3r committed:
Improve language in description of an input

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

dav3r committed:
Quote a variable name in an error message

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

dav3r committed:
Modify check of public egress tag value

This change aligns with the code that sets this tag in cisagov/cool-assessment-terraform (see https://github.com/cisagov/cool-assessment-terraform/pull/240).

dav3r committed:
Replace tuples with named tuples to improve code readability

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

dav3r committed:
Add error handling for user-provided static IPs in the config

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

mcdonnnj commented on README.md:
````diff
@@ -32,29 +31,11 @@ docker compose up build_deployment_package
 
 ```console
-docker compose down
+BUILD_FILE_NAME="publish_egress_ip_lambda.zip" docker compose up build_deployment_package
````
I'm not strongly against this, but wouldn't it make more sense to talk about configuring this in the Compose file?

dav3r replied:
I'm fine with mentioning both ways.  @mcdonnnj Can you write some verbiage about this in [skeleton-aws-lambda-python](https://github.com/cisagov/skeleton-aws-lambda-python) and then it can filter to all of the descendants?

mcdonnnj replied:
Good idea. I created https://github.com/cisagov/skeleton-aws-lambda-python/issues/26 to track this work.

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -1,24 +1,167 @@
+
+
+# Use a named tuple to hold EC2 information
+ec2_info = namedtuple("ec2_info", ["application_tag_value", "public_ip"])
```
Any issue with moving this up with the other `namedtuple` definition? It just feels strangely out of place here.

dav3r replied:
Please see https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/ff1698c71730998322b9b9161113196a02ebf46d.

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -36,54 +179,261 @@ def task_default(event):
-        contents: str = event.get("contents", "Hello from AWS Lambda!")
-        logger.info(
-            'Creating output using "%s" with contents "%s"', character, contents
-        )
-        result["message"] = cowsay.get_output_string(character, contents)
+    app_regex: re.Pattern[str]
+    description: str
+    filename: str
+    # I'd prefer to define ip_set as Set[Union[IPv4Address, IPv6Address]],
+    # but ipaddress.collapse_addresses() uses the type variable "_N" while
+    # ip_network returns Union[IPv4Network, IPv6network] which causes
+    # the mypy pre-commit hook to throw this error:
+    #   Value of type variable "_N" of "collapse_addresses" cannot be
+    #   "Union[IPv4Network, IPv6Network]"
+    # My solution to this problem is to simply define ip_set as Set[Any].
+    # For a similar issue and discussion, see
+    # https://github.com/python/typeshed/issues/2080
+    ip_set: Set[Any]
+    static_ips: List[str]
 
-    return result
+
+# Use a named tuple to store the results of the event validation
+event_validation = namedtuple("event_validation", ["event", "valid", "errors"])
```
Repeat of this feeling slightly out of place. I understand that these are stacked near a consuming function, but I'm more a fan of grouping this type of stuff vs. having it interspersed throughout.

dav3r replied:
I went back and forth on this.  I'm fine moving them all to the top and keeping them together.

dav3r replied:
Please see https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/ff1698c71730998322b9b9161113196a02ebf46d.

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -1,24 +1,167 @@
+"""AWS Lambda handler to publish egress IPs from a provided list of AWS accounts."""
 
 # Standard Python Libraries
+from collections import namedtuple
```
Thoughts on using [typing.NamedTuple](https://docs.python.org/3/library/typing.html#typing.NamedTuple) instead? This will allow us to bake typing information into the declaration.

dav3r replied:
Ask and ye shall receive: https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/f241c9c769b002c48ba56ff7a80746059568fb06

mcdonnnj reviewed (commented):
Some feedback items skimming through this again. Also a suggestion to update the lockfiles again for the final test pre-merge.

dav3r committed:
Group named tuple definitions together

dav3r committed:
Use typing.NamedTuple instead of collections.namedtuple

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -36,54 +203,257 @@ def task_default(event):
-def task_cowsay(event) -> dict[str, Union[Optional[str], bool]]:
-    """Generate an output message using the provided information."""
-    result: dict[str, Union[Optional[str], bool]] = {"message": None, "success": True}
+class FileConfig(TypedDict):
```
I would really like this moved up with the `NamedTuple` classes for consistency.

dav3r replied:
Your wish is my command: https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/47c12a984083bbdd4ce46067241ec459df64a2fb

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -1,24 +1,191 @@
+
+    # By default, new objects cannot be read by the public, but we want to
+    # allow public reads of this object
+    b_object.Acl().put(ACL="public-read")
```
The CloudFront configuration should negate the need for this functionality. Please see https://github.com/cisagov/cyhy_amis/commit/ad3435b8d7fbe170d3521fce9dfec36bd718a7f2 for more information.

```suggestion
```

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -36,54 +203,257 @@ def task_default(event):
+                errors.append(
+                    f'Invalid account ID(s) provided: "{", ".join(invalid_account_ids)}" - ID must be 12 digits.'
+                )
+        except Exception:
+            errors.append("account_ids must be a list of strings.")
```
Should this only apply to the initial `if`/`else` in this `except` block since it is `except`ing on the very broad `Exception`?

dav3r replied:
Good call, please see https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/8960c064c0b1d10e6bb4bb27d9823f6501c863d2.

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -36,54 +203,257 @@ def task_default(event):
+            domain=domain,
+            filename=config["filename"],
+            timestamp=now,
+            description=config["description"],
```
Since these are kwargs I think it would be useful to alphabetize them.

```suggestion
            description=config["description"],
            domain=domain,
            filename=config["filename"],
            timestamp=now,
```

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -1,24 +1,191 @@
 
-def failed_task(result: dict[str, Any], error_msg: str) -> None:
+# Define some named tuples to make the code more readable
+class aws_credentials(NamedTuple):
```
_Get ready to hate me anew_ but I think these NamedTuple definitions should follow typical naming conventions for classes.

dav3r replied:
CamelCase it is!  Please see https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/2861e2e56e296149945749e8baff0595267fce9b.

mcdonnnj replied:
Thank you but one nitpick is I don't think you intended to catch either of these in your replacement:
https://github.com/cisagov/publish-egress-ip-lambda/blob/2861e2e56e296149945749e8baff0595267fce9b/src/lambda_handler.py#L317
https://github.com/cisagov/publish-egress-ip-lambda/blob/2861e2e56e296149945749e8baff0595267fce9b/src/lambda_handler.py#L410

dav3r replied:
🤦  You are correct, good catch!  Cleaned up in https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/15e16531e71ce89bf306e838c29054855b76d506.

mcdonnnj reviewed (changes requested):
After more than just a skim I have some additional feedback for your consideration (sorry).

dav3r committed:
Relocate FileConfig class definition to be near other class definitions

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Prefer CamelCase for class names

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Alphabetize keyword arguments

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Remove public ACL configuration

Public access to items in the S3 bucket is configured in cisagov/publish-egress-ip-terraform.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Minimize try/except block

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Modify variable names unintentionally changed via an overzealous search and replace

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

mcdonnnj commented on Dockerfile:
```diff
@@ -31,13 +31,7 @@ FROM amazon/aws-lambda-python:$PY_VERSION as build-stage
-# person or group that maintains this repository; for example:
-# LABEL org.opencontainers.image.authors="vm-fusion-dev-group@trio.dhs.gov"
-LABEL org.opencontainers.image.authors="github@cisa.dhs.gov"
+LABEL org.opencontainers.image.authors="david.redmin@cisa.dhs.gov"
```
Just checking that you meant to use your `@cisa.dhs.gov` address here since we usually use our GWE addresses.

dav3r replied:
I'm fine with using my primary work email address here.

mcdonnnj commented on README.md:
```diff
@@ -2,13 +2,12 @@
-well as [pre-commit hooks](https://pre-commit.com) and
-[GitHub Actions](https://github.com/features/actions) configurations
-appropriate for the major languages that we use.
+This repository contains code to create a Lambda function that can scan a set
```
Since this repository technically only creates a Lambda function artifact that can be used to create an actual Lambda functioin.

```suggestion
This repository contains the code for a Lambda function that can scan a set
```

mcdonnnj commented on README.md:
```diff
@@ -72,12 +53,64 @@ cd src/py3.9
+| account\_ids | The list of AWS accounts to query for egress IPs to publish. | `list(string)` | n/a | yes |
+| application\_tag | The name of the AWS tag whose value represents the application associated with an IP address. | `string` | `"Application"` | no |
+| bucket\_name | The name of the S3 bucket to publish egress IP address information to. | `string` | n/a | yes |
+| domain | The domain hosting the published file(s) containing egress IPs. | `string` | `"example.gov"` | no |
```
Perhaps overkill but it might be nice to specify that it should be a FQDN as the default value could lead to misinterpretation.

dav3r replied:
Updated in https://github.com/cisagov/publish-egress-ip-lambda/pull/4/commits/ce708ebedc941aa5973450943a4e4cf6a503ba4b - let me know if that works.

mcdonnnj reviewed (approved):
Some minor items for your consideration but otherwise this LGTM.

dav3r committed:
Improve language in repo description

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Improve wording of an input description and make its default value match our example

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>
